# Functional Analysis for RNA-seq

```{r}
library(DESeq2)
library(tidyverse)
library(RColorBrewer)
library(pheatmap)
library(ggrepel)
library(cowplot)
library(clusterProfiler)
library(DEGreport)
library(org.Mm.eg.db)
library(DOSE)
library(pathview)
library(tximport)
library(AnnotationHub)
library(ensembldb)
library(apeglm)
```

## https://hbctraining.github.io/Intro-to-DGE/lessons/genomic_annotation.html
What we did in `step2_annotationhub.qmd` was create the ens2gene data frame from AnnotationHub.

However, for functional analysis, we also need Entrez IDs. So let's create a new data frame with Ensembl Gene IDs, Gene Symbols, and Entrez IDs.
```{r}
# Create a gene-level dataframe from mouse_ens (step2_annotationhub.qmd)
annotations_ahb <- genes(mouse_ens, return.type = "data.frame")  %>%
  dplyr::select(gene_id, gene_name, entrezid, gene_biotype) %>% 
  dplyr::filter(gene_id %in% res_table_shrunken_label_gene_tb$gene)
head(annotations_ahb)
nrow(annotations_ahb)
```

```{r}
# Wait a second, we don't have one-to-one mappings!
which(map(annotations_ahb$entrezid, length) > 1)
```

```{r}
# keep the first identifier for these multiple mapping cases
annotations_ahb$entrezid <- map(annotations_ahb$entrezid, 1) %>%  unlist()
head(annotations_ahb)
nrow(annotations_ahb)
```

```{r}
which(is.na(annotations_ahb$gene_name)) %>% length()
```

```{r}
which(duplicated(annotations_ahb$gene_name)) %>% length()
```

```{r}
# Determine the indices for the non-duplicated genes
non_duplicates_idx <- which(duplicated(annotations_ahb$gene_name) == FALSE)
length(non_duplicates_idx)
```

```{r}
# How many rows does annotations_ahb have?
annotations_ahb %>% nrow()

# Return only the non-duplicated genes using indices
annotations_ahb <- annotations_ahb[non_duplicates_idx, ]

# How many rows are we left with after removing?
annotations_ahb %>% nrow()
```

```{r}
# Determine how many of the Entrez column entries are NA
which(is.na(annotations_ahb$entrezid)) %>%  length()
```

## https://hbctraining.github.io/Intro-to-DGE/lessons/10_FA_over-representation_analysis.html

```{r}
# check previously processed data
head(res_table_shrunken_label_gene_tb)
nrow(res_table_shrunken_label_gene_tb)
```

```{r}
## Untested genes have padj = NA, so let's keep genes with padj != NA
res_table_shrunken_label_gene_noNAs_tb <- dplyr::filter(res_table_shrunken_label_gene_tb, !is.na(padj))
head(res_table_shrunken_label_gene_noNAs_tb)
nrow(res_table_shrunken_label_gene_noNAs_tb)
```

```{r}
## Merge the annotation with the results 
res_ids <- left_join(res_table_shrunken_label_gene_noNAs_tb, annotations_ahb, by=c("gene"="gene_id")) 
head(res_ids)
nrow(res_ids)
```

```{r}
## Create background dataset for hypergeometric testing using all tested genes for significance in the results                 
allOE_genes <- as.character(res_ids$gene)
head(allOE_genes)
```

```{r}
## Extract significant results
sigOE <- dplyr::filter(res_ids, padj < 0.05)
head(sigOE)
nrow(sigOE)
```

```{r}
# vector of significant over-expressed gene IDs
sigOE_genes <- as.character(sigOE$gene)
head(sigOE_genes)
```

```{r}
## Run GO enrichment analysis 
ego <- enrichGO(gene = sigOE_genes, 
                universe = allOE_genes,
                keyType = "ENSEMBL",
                OrgDb = org.Mm.eg.db, 
                ont = "BP", 
                pAdjustMethod = "BH", 
                qvalueCutoff = 0.05, 
                readable = TRUE)
head(ego)
```

```{r}
## Dotplot 
dotplot(ego, showCategory=15)
```

```{r}
p_dot_ego <- dotplot(ego, showCategory=15)
ggsave("plots/plot15_GO_dotplot.png", plot = p_dot_ego, width = 7, height = 7)
ggsave("plots/plot15_GO_dotplot.svg", plot = p_dot_ego, width = 7, height = 7)
```

```{r}
## Add similarity matrix to the termsim slot of enrichment result
ego <- enrichplot::pairwise_termsim(ego)
```

```{r}
## Enrichmap clusters the most significant (by padj) GO terms to visualize relationships between terms
emapplot(ego, showCategory = 15)
```

```{r}
## To color genes by log2 fold changes, we need to extract the log2 fold changes from our results table creating a named vector
OE_foldchanges <- sigOE$log2FoldChange
names(OE_foldchanges) <- sigOE$gene
head(OE_foldchanges)
```

```{r}
## Cnetplot details the genes associated with one or more terms - by default gives the top 5 significant terms (by padj)
cnetplot(ego, 
         showCategory = 1, 
         color.params=list(foldChange=OE_foldchanges),
         vertex.label.font=6)
```
         
```{r}
## If some of the high fold changes are getting drowned out due to a large range, you could set a maximum fold change value
OE_foldchanges <- ifelse(OE_foldchanges > 2, 2, OE_foldchanges)
OE_foldchanges <- ifelse(OE_foldchanges < -2, -2, OE_foldchanges)

cnetplot(ego, 
         categorySize="pvalue", 
         foldChange=OE_foldchanges, 
         showCategory = 1, 
         vertex.label.font=6)
```

```{r}
p_cnet_ego <- cnetplot(ego, 
         categorySize="pvalue", 
         foldChange=OE_foldchanges, 
         showCategory = 1, 
         vertex.label.font=6)
ggsave("plots/plot18_cnet_1cat_lfc.png", plot = p_cnet_ego, width = 7, height = 7)
ggsave("plots/plot18_cnet_1cat_lfc.svg", plot = p_cnet_ego, width = 7, height = 7)
```

## https://hbctraining.github.io/Intro-to-DGE/lessons/11_FA_functional_class_scoring.html

```{r}
## Remove any NA values (reduces the data by quite a bit)
res_entrez <- dplyr::filter(res_ids, entrezid != "NA")
head(res_entrez)
nrow(res_entrez)
```

```{r}
## Remove any Entrez duplicates
res_entrez <- res_entrez[which(duplicated(res_entrez$entrezid) == F), ]
head(res_entrez)
nrow(res_entrez)
```

```{r}
## Extract the foldchanges
foldchanges <- res_entrez$log2FoldChange

## Name each fold change with the corresponding Entrez ID
names(foldchanges) <- res_entrez$entrezid

## Sort fold changes in decreasing order
foldchanges <- sort(foldchanges, decreasing = TRUE)

head(foldchanges)
```

```{r}
set.seed(123456)

## GSEA using gene sets from KEGG pathways
gseaKEGG <- gseKEGG(geneList = foldchanges, # ordered named vector of fold changes (Entrez IDs are the associated names)
              organism = "mmu", # supported organisms listed below
              minGSSize = 20, # minimum gene set size (# genes in set) - change to test more sets or recover sets with fewer # genes
              pvalueCutoff = 0.05, # padj cutoff value
              verbose = FALSE)
class(gseaKEGG)
```

```{r}
## Extract the GSEA results
gseaKEGG_results <- gseaKEGG@result
head(gseaKEGG_results)
nrow(gseaKEGG_results)
```

```{r}
## Plot the GSEA plot for a single enriched pathway
gseaplot(gseaKEGG, geneSetID = 'mmu03030')
```

## https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html

```{r}
library(enrichplot)
```

```{r}
# upsetplot is an alternative to cnetplot for visualizing the complex association between genes and gene sets. It emphasizes the gene overlapping among different gene sets.
upsetplot(ego)
```

```{r}
# For GSEA result, it will plot the fold change distributions of different categories (e.g. unique to pathway, overlaps among different pathways).
upsetplot(gseaKEGG)
```

```{r}
ridgeplot(gseaKEGG, showCategory=15)
```

```{r}
# ridgeplot will visualize expression distributions of core enriched genes for GSEA enriched categories. It helps users to interpret up/down-regulated pathways.
p_ridge_kegg <- ridgeplot(gseaKEGG, showCategory=15)
ggsave("plots/plot22_ridge_kegg.png", plot = p_ridge_kegg, width = 7, height = 7)
ggsave("plots/plot22_ridge_kegg.svg", plot = p_ridge_kegg, width = 7, height = 7)
```

