<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.7.32">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>step6_func_analysis</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
html { -webkit-text-size-adjust: 100%; }
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="step6_func_analysis_files/libs/clipboard/clipboard.min.js"></script>
<script src="step6_func_analysis_files/libs/quarto-html/quarto.js" type="module"></script>
<script src="step6_func_analysis_files/libs/quarto-html/tabsets/tabsets.js" type="module"></script>
<script src="step6_func_analysis_files/libs/quarto-html/popper.min.js"></script>
<script src="step6_func_analysis_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="step6_func_analysis_files/libs/quarto-html/anchor.min.js"></script>
<link href="step6_func_analysis_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="step6_func_analysis_files/libs/quarto-html/quarto-syntax-highlighting-37eea08aefeeee20ff55810ff984fec1.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="step6_func_analysis_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="step6_func_analysis_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="step6_func_analysis_files/libs/bootstrap/bootstrap-81267100e462c21b3d6c0d5bf76a3417.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent quarto-light">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">




<section id="functional-analysis-for-rna-seq" class="level1">
<h1>Functional Analysis for RNA-seq</h1>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DESeq2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: S4Vectors</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: stats4</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiocGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'BiocGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    anyDuplicated, aperm, append, as.data.frame, basename, cbind,
    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,
    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,
    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,
    Position, rank, rbind, Reduce, rownames, sapply, saveRDS, setdiff,
    table, tapply, union, unique, unsplit, which.max, which.min</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'S4Vectors'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:utils':

    findMatches</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:base':

    expand.grid, I, unname</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: IRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'IRanges' was built under R version 4.4.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomicRanges</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomeInfoDb</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'GenomeInfoDb' was built under R version 4.4.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: SummarizedExperiment</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: MatrixGenerics</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'MatrixGenerics' was built under R version 4.4.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: matrixStats</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'MatrixGenerics'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:matrixStats':

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: Biobase</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'Biobase'</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following object is masked from 'package:MatrixGenerics':

    rowMedians</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>The following objects are masked from 'package:matrixStats':

    anyMissing, rowMedians</code></pre>
</div>
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'ggplot2' was built under R version 4.4.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'tidyr' was built under R version 4.4.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'readr' was built under R version 4.4.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.6
✔ forcats   1.0.1     ✔ stringr   1.6.0
✔ ggplot2   4.0.1     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.2
✔ purrr     1.2.0     </code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ lubridate::%within%() masks IRanges::%within%()
✖ dplyr::collapse()     masks IRanges::collapse()
✖ dplyr::combine()      masks Biobase::combine(), BiocGenerics::combine()
✖ dplyr::count()        masks matrixStats::count()
✖ dplyr::desc()         masks IRanges::desc()
✖ tidyr::expand()       masks S4Vectors::expand()
✖ dplyr::filter()       masks stats::filter()
✖ dplyr::first()        masks S4Vectors::first()
✖ dplyr::lag()          masks stats::lag()
✖ ggplot2::Position()   masks BiocGenerics::Position(), base::Position()
✖ purrr::reduce()       masks GenomicRanges::reduce(), IRanges::reduce()
✖ dplyr::rename()       masks S4Vectors::rename()
✖ lubridate::second()   masks S4Vectors::second()
✖ lubridate::second&lt;-() masks S4Vectors::second&lt;-()
✖ dplyr::slice()        masks IRanges::slice()
ℹ Use the conflicted package (&lt;http://conflicted.r-lib.org/&gt;) to force all conflicts to become errors</code></pre>
</div>
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pheatmap)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggrepel)</span>
<span id="cb33-4"><a href="#cb33-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(cowplot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>
Attaching package: 'cowplot'

The following object is masked from 'package:lubridate':

    stamp</code></pre>
</div>
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(clusterProfiler)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'clusterProfiler' was built under R version 4.4.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
clusterProfiler v4.14.6 Learn more at https://yulab-smu.top/contribution-knowledge-mining/

Please cite:

Guangchuang Yu, Li-Gen Wang, Yanyan Han and Qing-Yu He.
clusterProfiler: an R package for comparing biological themes among
gene clusters. OMICS: A Journal of Integrative Biology. 2012,
16(5):284-287

Attaching package: 'clusterProfiler'

The following object is masked from 'package:purrr':

    simplify

The following object is masked from 'package:IRanges':

    slice

The following object is masked from 'package:S4Vectors':

    rename

The following object is masked from 'package:stats':

    filter</code></pre>
</div>
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DEGreport)</span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(org.Mm.eg.db)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: AnnotationDbi

Attaching package: 'AnnotationDbi'

The following object is masked from 'package:clusterProfiler':

    select

The following object is masked from 'package:dplyr':

    select</code></pre>
</div>
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(DOSE)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'DOSE' was built under R version 4.4.3</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>DOSE v4.0.1 Learn more at https://yulab-smu.top/contribution-knowledge-mining/

Please cite:

Guangchuang Yu, Li-Gen Wang, Guang-Rong Yan, Qing-Yu He. DOSE: an
R/Bioconductor package for Disease Ontology Semantic and Enrichment
analysis. Bioinformatics. 2015, 31(4):608-609</code></pre>
</div>
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pathview)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'pathview' was built under R version 4.4.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>
##############################################################################
Pathview is an open source software package distributed under GNU General
Public License version 3 (GPLv3). Details of GPLv3 is available at
http://www.gnu.org/licenses/gpl-3.0.html. Particullary, users are required to
formally cite the original Pathview paper (not just mention it) in publications
or products. For details, do citation("pathview") within R.

The pathview downloads and uses KEGG data. Non-academic uses may require a KEGG
license agreement (details at http://www.kegg.jp/kegg/legal.html).
##############################################################################</code></pre>
</div>
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tximport)</span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(AnnotationHub)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: BiocFileCache
Loading required package: dbplyr

Attaching package: 'dbplyr'

The following objects are masked from 'package:dplyr':

    ident, sql


Attaching package: 'AnnotationHub'

The following object is masked from 'package:Biobase':

    cache</code></pre>
</div>
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ensembldb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Loading required package: GenomicFeatures
Loading required package: AnnotationFilter

Attaching package: 'ensembldb'

The following object is masked from 'package:clusterProfiler':

    filter

The following object is masked from 'package:dplyr':

    filter

The following object is masked from 'package:stats':

    filter</code></pre>
</div>
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(apeglm)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="httpshbctraining.github.iointro-to-dgelessonsgenomic_annotation.html" class="level2">
<h2 class="anchored" data-anchor-id="httpshbctraining.github.iointro-to-dgelessonsgenomic_annotation.html">https://hbctraining.github.io/Intro-to-DGE/lessons/genomic_annotation.html</h2>
<p>What we did in <code>step2_annotationhub.qmd</code> was create the ens2gene data frame from AnnotationHub.</p>
<p>However, for functional analysis, we also need Entrez IDs. So let’s create a new data frame with Ensembl Gene IDs, Gene Symbols, and Entrez IDs.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load objects from previous steps</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a><span class="fu">load</span>(<span class="st">"intermediate/step5_objects.RData"</span>)  <span class="co"># res_table_shrunken_label_gene_tb, ens2gene</span></span>
<span id="cb51-3"><a href="#cb51-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb51-4"><a href="#cb51-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Load mouse_ens from step2</span></span>
<span id="cb51-5"><a href="#cb51-5" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(AnnotationHub)</span>
<span id="cb51-6"><a href="#cb51-6" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ensembldb)</span>
<span id="cb51-7"><a href="#cb51-7" aria-hidden="true" tabindex="-1"></a>ah <span class="ot">&lt;-</span> <span class="fu">AnnotationHub</span>()</span>
<span id="cb51-8"><a href="#cb51-8" aria-hidden="true" tabindex="-1"></a>mouse_ens <span class="ot">&lt;-</span> <span class="fu">query</span>(ah, <span class="fu">c</span>(<span class="st">"Mus musculus"</span>, <span class="st">"EnsDb"</span>))[[<span class="st">"AH119358"</span>]]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>loading from cache</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a gene-level dataframe from mouse_ens (step2_annotationhub.qmd)</span></span>
<span id="cb53-2"><a href="#cb53-2" aria-hidden="true" tabindex="-1"></a>annotations_ahb <span class="ot">&lt;-</span> <span class="fu">genes</span>(mouse_ens, <span class="at">return.type =</span> <span class="st">"data.frame"</span>)  <span class="sc">%&gt;%</span></span>
<span id="cb53-3"><a href="#cb53-3" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">select</span>(gene_id, gene_name, entrezid, gene_biotype) <span class="sc">%&gt;%</span> </span>
<span id="cb53-4"><a href="#cb53-4" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">filter</span>(gene_id <span class="sc">%in%</span> res_table_shrunken_label_gene_tb<span class="sc">$</span>gene)</span>
<span id="cb53-5"><a href="#cb53-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(annotations_ahb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             gene_id     gene_name  entrezid         gene_biotype
1 ENSMUSG00000102693 4933401J01Rik        NA                  TEC
2 ENSMUSG00000064842       Gm26206 115487594                snRNA
3 ENSMUSG00000051951          Xkr4    497097       protein_coding
4 ENSMUSG00000102851       Gm18956        NA processed_pseudogene
5 ENSMUSG00000103377       Gm37180        NA                  TEC
6 ENSMUSG00000104017       Gm37363        NA                  TEC</code></pre>
</div>
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(annotations_ahb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 55613</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Wait a second, we don't have one-to-one mappings!</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(<span class="fu">map</span>(annotations_ahb<span class="sc">$</span>entrezid, length) <span class="sc">&gt;</span> <span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ENSMUSG00000086727 ENSMUSG00000100553 ENSMUSG00000101483 ENSMUSG00000044783 
               402               1298               1350               1582 
ENSMUSG00000100457 ENSMUSG00000070645 ENSMUSG00000038209 ENSMUSG00000019868 
              1887               2218               2977               3701 
ENSMUSG00000019865 ENSMUSG00000078452 ENSMUSG00000112376 ENSMUSG00000075046 
              3703               3836               4382               4441 
ENSMUSG00000037747 ENSMUSG00000052291 ENSMUSG00000112627 ENSMUSG00000112846 
              4663               5725               5736               5849 
ENSMUSG00000040195 ENSMUSG00000095075 ENSMUSG00000085588 ENSMUSG00000119011 
              5980               6104               6724               6958 
ENSMUSG00000119662 ENSMUSG00000085528 ENSMUSG00000107877 ENSMUSG00000058546 
              7484               7609               7616               7849 
ENSMUSG00000093938 ENSMUSG00000119476 ENSMUSG00000018378 ENSMUSG00000087259 
              7903               8112               8146               8912 
ENSMUSG00000062352 ENSMUSG00000073164 ENSMUSG00000094002 ENSMUSG00000079036 
              9610               9651               9733              10746 
ENSMUSG00000056457 ENSMUSG00000021557 ENSMUSG00000050334 ENSMUSG00000078941 
             11951              13090              13681              13960 
ENSMUSG00000023156 ENSMUSG00000094132 ENSMUSG00000057606 ENSMUSG00000091477 
             14420              14531              14854              15118 
ENSMUSG00000053961 ENSMUSG00000121896 ENSMUSG00000091122 ENSMUSG00000021846 
             15153              15165              15196              15323 
ENSMUSG00000050766 ENSMUSG00000115422 ENSMUSG00000041841 ENSMUSG00000115230 
             15444              16349              16806              17266 
ENSMUSG00000121583 ENSMUSG00000047347 ENSMUSG00000062037 ENSMUSG00000036925 
             17771              18059              18377              18680 
ENSMUSG00000005983 ENSMUSG00000008658 ENSMUSG00000109857 ENSMUSG00000121816 
             18707              18765              18862              19226 
ENSMUSG00000022820 ENSMUSG00000063137 ENSMUSG00000064006 ENSMUSG00000095706 
             19494              19762              19775              19792 
ENSMUSG00000100857 ENSMUSG00000068071 ENSMUSG00000068068 ENSMUSG00000023809 
             19933              20094              20097              20397 
ENSMUSG00000062327 ENSMUSG00000002250 ENSMUSG00000060550 ENSMUSG00000117190 
             20427              21086              21509              21790 
ENSMUSG00000117011 ENSMUSG00000063087 ENSMUSG00000117473 ENSMUSG00000059040 
             21816              22743              23021              23466 
ENSMUSG00000086607 ENSMUSG00000089156 ENSMUSG00000024571 ENSMUSG00000056629 
             23569              23668              23958              24388 
ENSMUSG00000003680 ENSMUSG00000062199 ENSMUSG00000053279 ENSMUSG00000035171 
             24462              24688              24800              24814 
ENSMUSG00000097740 ENSMUSG00000071528 ENSMUSG00000027022 ENSMUSG00000052371 
             25164              25346              26799              26956 
ENSMUSG00000054510 ENSMUSG00000075061 ENSMUSG00000089859 ENSMUSG00000086462 
             27019              27461              27463              27867 
ENSMUSG00000001768 ENSMUSG00000087382 ENSMUSG00000078899 ENSMUSG00000102699 
             28411              29117              29179              29538 
ENSMUSG00000106441 ENSMUSG00000036381 ENSMUSG00000063767 ENSMUSG00000000562 
             29827              30151              30749              31329 
ENSMUSG00000096770 ENSMUSG00000062006 ENSMUSG00000105923 ENSMUSG00000115466 
             31498              31799              32014              32262 
ENSMUSG00000073888 ENSMUSG00000078746 ENSMUSG00000073877 ENSMUSG00000093996 
             32730              32738              32748              32763 
ENSMUSG00000089717 ENSMUSG00000028700 ENSMUSG00000047675 ENSMUSG00000057021 
             32990              33893              33941              35163 
ENSMUSG00000029089 ENSMUSG00000107232 ENSMUSG00000074011 ENSMUSG00000092166 
             36067              36578              36846              36860 
ENSMUSG00000029592 ENSMUSG00000072694 ENSMUSG00000085708 ENSMUSG00000004455 
             37327              37351              37496              37588 
ENSMUSG00000029723 ENSMUSG00000097321 ENSMUSG00000045095 ENSMUSG00000072872 
             38037              38461              40437              40510 
ENSMUSG00000030337 ENSMUSG00000079852 ENSMUSG00000070995 ENSMUSG00000074419 
             41027              41215              41628              41721 
ENSMUSG00000074417 ENSMUSG00000092456 ENSMUSG00000097972 ENSMUSG00000098094 
             41725              42665              43152              43161 
ENSMUSG00000060565 ENSMUSG00000053742 ENSMUSG00000074141 ENSMUSG00000003271 
             43311              43336              43592              43692 
ENSMUSG00000059146 ENSMUSG00000097016 ENSMUSG00000059319 ENSMUSG00000048469 
             44325              44511              44592              44957 
ENSMUSG00000052305 ENSMUSG00000052749 ENSMUSG00000000486 ENSMUSG00000109291 
             45037              45086              45739              46455 
ENSMUSG00000015341 ENSMUSG00000079058 ENSMUSG00000079033 ENSMUSG00000087408 
             46794              47122              47510              47518 
ENSMUSG00000003812 ENSMUSG00000119783 ENSMUSG00000049864 ENSMUSG00000096757 
             47885              49387              49516              49517 
ENSMUSG00000032872 ENSMUSG00000062270 ENSMUSG00001074846 ENSMUSG00000025646 
             50727              50802              51155              51307 
ENSMUSG00000091537 ENSMUSG00000094383 ENSMUSG00000094558 ENSMUSG00000060685 
             51309              51746              51802              51958 
ENSMUSG00000087358 ENSMUSG00000096468 ENSMUSG00000094714 ENSMUSG00000015290 
             52136              52575              52576              52908 
ENSMUSG00000032750 ENSMUSG00000086008 ENSMUSG00000095950 ENSMUSG00000094813 
             52944              54352              54602              54611 
ENSMUSG00000104267 ENSMUSG00000096820 ENSMUSG00000093903 ENSMUSG00000095900 
             54635              54636              54638              54645 
ENSMUSG00000095365 ENSMUSG00000094325 ENSMUSG00000102739 ENSMUSG00000095267 
             54685              54693              54694              54699 
ENSMUSG00000093868 ENSMUSG00000101158 ENSMUSG00000095793 ENSMUSG00000094575 
             54785              54819              54889              54935 
ENSMUSG00000095452 ENSMUSG00000100485 ENSMUSG00000094729 ENSMUSG00000101155 
             54979              55078              55095              55147 
ENSMUSG00000096122 ENSMUSG00000094821 ENSMUSG00000093950 ENSMUSG00000101766 
             55156              55220              55223              55242 
ENSMUSG00000095785 ENSMUSG00000096666 ENSMUSG00000095302 ENSMUSG00000096016 
             55250              55261              55264              55278 
ENSMUSG00000095634 ENSMUSG00000094052 ENSMUSG00000096902 ENSMUSG00000094647 
             55283              55309              55342              55419 
ENSMUSG00000096898 ENSMUSG00000094739 ENSMUSG00000095867 ENSMUSG00000094746 
             55424              55441              55445              55447 
ENSMUSG00000101286 
             55480 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># keep the first identifier for these multiple mapping cases</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a>annotations_ahb<span class="sc">$</span>entrezid <span class="ot">&lt;-</span> <span class="fu">map</span>(annotations_ahb<span class="sc">$</span>entrezid, <span class="dv">1</span>) <span class="sc">%&gt;%</span>  <span class="fu">unlist</span>()</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(annotations_ahb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>             gene_id     gene_name  entrezid         gene_biotype
1 ENSMUSG00000102693 4933401J01Rik        NA                  TEC
2 ENSMUSG00000064842       Gm26206 115487594                snRNA
3 ENSMUSG00000051951          Xkr4    497097       protein_coding
4 ENSMUSG00000102851       Gm18956        NA processed_pseudogene
5 ENSMUSG00000103377       Gm37180        NA                  TEC
6 ENSMUSG00000104017       Gm37363        NA                  TEC</code></pre>
</div>
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(annotations_ahb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 55613</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(<span class="fu">is.na</span>(annotations_ahb<span class="sc">$</span>gene_name)) <span class="sc">%&gt;%</span> <span class="fu">length</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 0</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(<span class="fu">duplicated</span>(annotations_ahb<span class="sc">$</span>gene_name)) <span class="sc">%&gt;%</span> <span class="fu">length</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 677</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine the indices for the non-duplicated genes</span></span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>non_duplicates_idx <span class="ot">&lt;-</span> <span class="fu">which</span>(<span class="fu">duplicated</span>(annotations_ahb<span class="sc">$</span>gene_name) <span class="sc">==</span> <span class="cn">FALSE</span>)</span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a><span class="fu">length</span>(non_duplicates_idx)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 54936</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># How many rows does annotations_ahb have?</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>annotations_ahb <span class="sc">%&gt;%</span> <span class="fu">nrow</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 55613</code></pre>
</div>
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Return only the non-duplicated genes using indices</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a>annotations_ahb <span class="ot">&lt;-</span> annotations_ahb[non_duplicates_idx, ]</span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a><span class="co"># How many rows are we left with after removing?</span></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>annotations_ahb <span class="sc">%&gt;%</span> <span class="fu">nrow</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 54936</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine how many of the Entrez column entries are NA</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a><span class="fu">which</span>(<span class="fu">is.na</span>(annotations_ahb<span class="sc">$</span>entrezid)) <span class="sc">%&gt;%</span>  <span class="fu">length</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 27965</code></pre>
</div>
</div>
</section>
<section id="httpshbctraining.github.iointro-to-dgelessons10_fa_over-representation_analysis.html" class="level2">
<h2 class="anchored" data-anchor-id="httpshbctraining.github.iointro-to-dgelessons10_fa_over-representation_analysis.html">https://hbctraining.github.io/Intro-to-DGE/lessons/10_FA_over-representation_analysis.html</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a><span class="co"># check previously processed data</span></span>
<span id="cb75-2"><a href="#cb75-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_table_shrunken_label_gene_tb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 9
  gene   baseMean log2FoldChange  lfcSE pvalue  padj threshold symbol genelabels
  &lt;chr&gt;     &lt;dbl&gt;          &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;lgl&gt;     &lt;chr&gt;  &lt;chr&gt;     
1 ENSMU…    2198.           2.49 0.0625      0     0 TRUE      Nhp2   Nhp2      
2 ENSMU…    1186.           3.20 0.0829      0     0 TRUE      Smc2   Smc2      
3 ENSMU…    1362.           4.66 0.0951      0     0 TRUE      Kif20a Kif20a    
4 ENSMU…    1272.           2.93 0.0748      0     0 TRUE      Srpk1  Srpk1     
5 ENSMU…    4374.           4.21 0.0917      0     0 TRUE      Nolc1  Nolc1     
6 ENSMU…    3387.           2.03 0.0537      0     0 TRUE      Srsf6  Srsf6     </code></pre>
</div>
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(res_table_shrunken_label_gene_tb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 55613</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Untested genes have padj = NA, so let's keep genes with padj != NA</span></span>
<span id="cb79-2"><a href="#cb79-2" aria-hidden="true" tabindex="-1"></a>res_table_shrunken_label_gene_noNAs_tb <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(res_table_shrunken_label_gene_tb, <span class="sc">!</span><span class="fu">is.na</span>(padj))</span>
<span id="cb79-3"><a href="#cb79-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_table_shrunken_label_gene_noNAs_tb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 9
  gene   baseMean log2FoldChange  lfcSE pvalue  padj threshold symbol genelabels
  &lt;chr&gt;     &lt;dbl&gt;          &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;lgl&gt;     &lt;chr&gt;  &lt;chr&gt;     
1 ENSMU…    2198.           2.49 0.0625      0     0 TRUE      Nhp2   Nhp2      
2 ENSMU…    1186.           3.20 0.0829      0     0 TRUE      Smc2   Smc2      
3 ENSMU…    1362.           4.66 0.0951      0     0 TRUE      Kif20a Kif20a    
4 ENSMU…    1272.           2.93 0.0748      0     0 TRUE      Srpk1  Srpk1     
5 ENSMU…    4374.           4.21 0.0917      0     0 TRUE      Nolc1  Nolc1     
6 ENSMU…    3387.           2.03 0.0537      0     0 TRUE      Srsf6  Srsf6     </code></pre>
</div>
<div class="sourceCode cell-code" id="cb81"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb81-1"><a href="#cb81-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(res_table_shrunken_label_gene_noNAs_tb)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 19789</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb83"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb83-1"><a href="#cb83-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Merge the annotation with the results </span></span>
<span id="cb83-2"><a href="#cb83-2" aria-hidden="true" tabindex="-1"></a>res_ids <span class="ot">&lt;-</span> <span class="fu">left_join</span>(res_table_shrunken_label_gene_noNAs_tb, annotations_ahb, <span class="at">by=</span><span class="fu">c</span>(<span class="st">"gene"</span><span class="ot">=</span><span class="st">"gene_id"</span>)) </span>
<span id="cb83-3"><a href="#cb83-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_ids)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 12
  gene   baseMean log2FoldChange  lfcSE pvalue  padj threshold symbol genelabels
  &lt;chr&gt;     &lt;dbl&gt;          &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;lgl&gt;     &lt;chr&gt;  &lt;chr&gt;     
1 ENSMU…    2198.           2.49 0.0625      0     0 TRUE      Nhp2   Nhp2      
2 ENSMU…    1186.           3.20 0.0829      0     0 TRUE      Smc2   Smc2      
3 ENSMU…    1362.           4.66 0.0951      0     0 TRUE      Kif20a Kif20a    
4 ENSMU…    1272.           2.93 0.0748      0     0 TRUE      Srpk1  Srpk1     
5 ENSMU…    4374.           4.21 0.0917      0     0 TRUE      Nolc1  Nolc1     
6 ENSMU…    3387.           2.03 0.0537      0     0 TRUE      Srsf6  Srsf6     
# ℹ 3 more variables: gene_name &lt;chr&gt;, entrezid &lt;int&gt;, gene_biotype &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb85"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb85-1"><a href="#cb85-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(res_ids)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 19789</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb87"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb87-1"><a href="#cb87-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Create background dataset for hypergeometric testing using all tested genes for significance in the results                 </span></span>
<span id="cb87-2"><a href="#cb87-2" aria-hidden="true" tabindex="-1"></a>allOE_genes <span class="ot">&lt;-</span> <span class="fu">as.character</span>(res_ids<span class="sc">$</span>gene)</span>
<span id="cb87-3"><a href="#cb87-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(allOE_genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ENSMUSG00000001056" "ENSMUSG00000028312" "ENSMUSG00000003779"
[4] "ENSMUSG00000004865" "ENSMUSG00000015176" "ENSMUSG00000016921"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb89"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb89-1"><a href="#cb89-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Extract significant results</span></span>
<span id="cb89-2"><a href="#cb89-2" aria-hidden="true" tabindex="-1"></a>sigOE <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(res_ids, padj <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb89-3"><a href="#cb89-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sigOE)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 12
  gene   baseMean log2FoldChange  lfcSE pvalue  padj threshold symbol genelabels
  &lt;chr&gt;     &lt;dbl&gt;          &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;lgl&gt;     &lt;chr&gt;  &lt;chr&gt;     
1 ENSMU…    2198.           2.49 0.0625      0     0 TRUE      Nhp2   Nhp2      
2 ENSMU…    1186.           3.20 0.0829      0     0 TRUE      Smc2   Smc2      
3 ENSMU…    1362.           4.66 0.0951      0     0 TRUE      Kif20a Kif20a    
4 ENSMU…    1272.           2.93 0.0748      0     0 TRUE      Srpk1  Srpk1     
5 ENSMU…    4374.           4.21 0.0917      0     0 TRUE      Nolc1  Nolc1     
6 ENSMU…    3387.           2.03 0.0537      0     0 TRUE      Srsf6  Srsf6     
# ℹ 3 more variables: gene_name &lt;chr&gt;, entrezid &lt;int&gt;, gene_biotype &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb91"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb91-1"><a href="#cb91-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(sigOE)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 12612</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb93"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb93-1"><a href="#cb93-1" aria-hidden="true" tabindex="-1"></a><span class="co"># vector of significant over-expressed gene IDs</span></span>
<span id="cb93-2"><a href="#cb93-2" aria-hidden="true" tabindex="-1"></a>sigOE_genes <span class="ot">&lt;-</span> <span class="fu">as.character</span>(sigOE<span class="sc">$</span>gene)</span>
<span id="cb93-3"><a href="#cb93-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(sigOE_genes)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "ENSMUSG00000001056" "ENSMUSG00000028312" "ENSMUSG00000003779"
[4] "ENSMUSG00000004865" "ENSMUSG00000015176" "ENSMUSG00000016921"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb95"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb95-1"><a href="#cb95-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Run GO enrichment analysis </span></span>
<span id="cb95-2"><a href="#cb95-2" aria-hidden="true" tabindex="-1"></a>ego <span class="ot">&lt;-</span> <span class="fu">enrichGO</span>(<span class="at">gene =</span> sigOE_genes, </span>
<span id="cb95-3"><a href="#cb95-3" aria-hidden="true" tabindex="-1"></a>                <span class="at">universe =</span> allOE_genes,</span>
<span id="cb95-4"><a href="#cb95-4" aria-hidden="true" tabindex="-1"></a>                <span class="at">keyType =</span> <span class="st">"ENSEMBL"</span>,</span>
<span id="cb95-5"><a href="#cb95-5" aria-hidden="true" tabindex="-1"></a>                <span class="at">OrgDb =</span> org.Mm.eg.db, </span>
<span id="cb95-6"><a href="#cb95-6" aria-hidden="true" tabindex="-1"></a>                <span class="at">ont =</span> <span class="st">"BP"</span>, </span>
<span id="cb95-7"><a href="#cb95-7" aria-hidden="true" tabindex="-1"></a>                <span class="at">pAdjustMethod =</span> <span class="st">"BH"</span>, </span>
<span id="cb95-8"><a href="#cb95-8" aria-hidden="true" tabindex="-1"></a>                <span class="at">qvalueCutoff =</span> <span class="fl">0.05</span>, </span>
<span id="cb95-9"><a href="#cb95-9" aria-hidden="true" tabindex="-1"></a>                <span class="at">readable =</span> <span class="cn">TRUE</span>)</span>
<span id="cb95-10"><a href="#cb95-10" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(ego)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                   ID                          Description GeneRatio   BgRatio
GO:0022613 GO:0022613 ribonucleoprotein complex biogenesis 401/11044 437/15389
GO:0042254 GO:0042254                  ribosome biogenesis 283/11044 310/15389
GO:0006397 GO:0006397                      mRNA processing 397/11044 463/15389
GO:0006364 GO:0006364                      rRNA processing 196/11044 215/15389
GO:0022618 GO:0022618         protein-RNA complex assembly 164/11044 177/15389
GO:0016072 GO:0016072               rRNA metabolic process 223/11044 249/15389
           RichFactor FoldEnrichment   zScore       pvalue     p.adjust
GO:0022613  0.9176201       1.278636 9.420799 7.234824e-26 4.374898e-22
GO:0042254  0.9129032       1.272063 7.714789 9.492590e-18 2.870085e-14
GO:0006397  0.8574514       1.194795 6.785107 3.634529e-13 7.325998e-10
GO:0006364  0.9116279       1.270286 6.362864 1.573231e-12 2.378332e-09
GO:0022618  0.9265537       1.291084 6.209714 2.901178e-12 3.508685e-09
GO:0016072  0.8955823       1.247928 6.288132 5.100368e-12 5.140321e-09
                 qvalue
GO:0022613 3.717938e-22
GO:0042254 2.439096e-14
GO:0006397 6.225884e-10
GO:0006364 2.021188e-09
GO:0022618 2.981801e-09
GO:0016072 4.368420e-09
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               geneID
GO:0022613                                                              Nhp2/Srpk1/Srsf6/Ddx27/C1qbp/Nop16/Ncl/Ddx20/Wdr3/Npm1/Nop2/Ddx21/Utp4/Pdcd11/Ddx18/Bms1/Wdr77/Eif5/Nop56/Mtrex/Urb2/Lyar/Noc4l/Wdr46/Wdr12/Nol8/Dnttip2/Ftsj3/Nop14/Eif3j2/Utp15/Nle1/Rps7/Pa2g4/Eif3g/Prpf3/Riok1/Nop10/Sf3a3/Ruvbl1/Esf1/Wdr55/Ppan/Grwd1/Denr/Imp3/Exosc3/Eif6/Gtf3a/Tsr1/Mettl16/Xpo1/Bop1/Gar1/Ebna1bp2/Nol10/Prpf4b/Taf9/Dcaf13/Gemin6/Mphosph10/Suv39h1/Pwp2/Brix1/Wdr43/Sdad1/Rps4x/2810004N23Rik/Exosc2/Eif3b/Nmd3/Rrp1b/Rplp0/Ddx10/Ran/Riox1/Rcl1/Ddx3x/Exosc1/Nat10/Ago1/Mybbp1a/Fbl/Nop58/Luc7l2/Utp20/Glul/Snrpb/Eif2s2/Mrps2/Wdr75/Lsm2/Wdr74/Pak1ip1/Prmt7/Rps3a1/Eri1/Gnl2/Dis3/Aatf/Prpf19/U2af2/Gemin5/Nip7/Rpusd2/Rpf1/Tarbp2/Dkc1/Gtpbp4/Heatr3/Emg1/Cul4b/Sf3a2/Rbm34/Scaf11/Snrpd3/Eif3a/Mrto4/Prpf31/Ddx56/Urb1/Krr1/Nol9/Rexo4/Rpp40/Bud23/Exosc7/Smn1/Npm3/Atr/Dhx37/Fcf1/Nsun5/Eif3l/Rrs1/Utp14a/Wdr36/Eif3e/Ngdn/Rps13/Rps6/Cinp/Rpp30/Eif2a/Mterf3/Dhx9/Srsf1/Aar2/Utp18/Rrp15/Rps5/Exosc5/Snrnp200/Tma16/Naf1/Sf3b1/Sf3a1/Rpl14/Ago4/Rpl35/Isy1/Eif4a3/Pno1/Exosc10/Ncbp1/Mcts2/Rps19bp1/Hsp90ab1/Rrp9/Hsp90aa1/Ddx52/Shq1/Tbl3/Mak16/Airim/Snrpe/Eif3j1/Mterf4/Ddx47/Noc2l/Surf6/Mrm3/Prmt5/Sart3/Pop7/Riok2/Eri3/Gemin2/Rps21/Rps8/Rpl7l1/Tsr2/Rpl7/Drosha/Nufip1/Rpsa/Srfbp1/Usp39/Rrp8/Ltv1/Rps17/Exosc8/Rpusd4/Utp3/Ythdc1/Rps16/Utp11/Znhit3/Eif2s3y/Riox2/Fastkd2/Srsf10/Rcc1l/Nol3/Prpf6/Clp1/Nup88/Rps15a/Rpp38/Eif3i/Nudt16/Ddx54/Ruvbl2/Ddx51/Dhx29/Isg20l2/Cdkn2a/Utp25/Rps27rt/Nopchap1/Dimt1/Lsg1/Nvl/Pes1/Imp4/Pelp1/Tsc1/Luc7l3/Mrm2/Eri2/Heatr1/Nol6/Rps9/Eif3c/Lsm4/Eif3m/Setx/Mphosph6/Snrpd1/Slx9/Khdc4/Noa1/Snrpf/Gemin4/Bysl/Eif3d/Rps27a/Eif3h/Usp4/Nol7/Zfp622/Sart1/Xab2/Celf2/Ercc2/Mrpl10/Cul4a/Prpf39/Rpl35a/Rpl27/Celf5/Snrpg/Afg2b/Nom1/Nsa2/Rps28/Mettl17/Nob1/Wdr18/Rps23/Ddx31/Nol11/Znhit6/Eif2s3x/Atm/Las1l/Slu7/Tfb2m/Rpf2/Snrpd2/Rsrp1/Prpf8/Pih1d1/Abt1/Rps12/Utp6/Adar/Pin4/Isg20/Utp14b/Snrpc/Clns1a/Mettl18/Usp16/Mpv17l2/Zfp593/Rbm10/Rsl24d1/Rpl7a/Ckap5/Yju2/Rpl11/Mbnl1/Mdn1/Rbfa/Celf3/Snu13/Rps25/Rps24/Efl1/Usp36/Prkra/Tssc4/Sde2/Ddx49/Fbll1/Utp23/Gcfc2/Prkdc/Pop5/Zrsr2/Prpf18/Rpl38/Strap/Zcchc4/Rpl5/Ipo9/Luc7l/Crnkl1/Rpl24/Ptges3/Rps19/Sfswap/Celf1/Celf6/Eif1a/Wbp11/Mrm1/Srpk3/Kri1/Pwp1/Rpl26/Ddx39b/Tfb1m/Ddx46/Mrps7/Rps27l/Ythdf2/Frg1/Mettl15/Tent4b/Pten/Rps15/Rps27/Rps11/Eif5b/Nop9/LTO1/Fau/Ddx28/Eral1/Lsm6/Eif2d/Rrp1/Pop4/Rrn3/Nsun3/Mettl5/Ago3/Pih1d2/Afg2a/Rpusd1/Tsr3/Riok3/Coil/Ago2/Rpl13a/Htatsf1/Rexo5/Rrp7a/Rps14/Rbm5/Xrcc5/Gemin7/Rexo1/Eif3k/Trmt2b/Mcat/Eif3f/Srsf12/Puf60
GO:0042254                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Nhp2/Ddx27/C1qbp/Nop16/Ncl/Wdr3/Npm1/Nop2/Ddx21/Utp4/Pdcd11/Ddx18/Bms1/Eif5/Nop56/Mtrex/Urb2/Lyar/Noc4l/Wdr46/Wdr12/Nol8/Dnttip2/Ftsj3/Nop14/Utp15/Nle1/Rps7/Pa2g4/Riok1/Nop10/Esf1/Wdr55/Ppan/Grwd1/Imp3/Exosc3/Eif6/Gtf3a/Tsr1/Mettl16/Xpo1/Bop1/Gar1/Ebna1bp2/Nol10/Taf9/Dcaf13/Mphosph10/Suv39h1/Pwp2/Brix1/Wdr43/Sdad1/Rps4x/2810004N23Rik/Exosc2/Nmd3/Rrp1b/Rplp0/Ddx10/Ran/Riox1/Rcl1/Ddx3x/Exosc1/Nat10/Mybbp1a/Fbl/Nop58/Utp20/Glul/Mrps2/Wdr75/Wdr74/Pak1ip1/Rps3a1/Eri1/Gnl2/Dis3/Aatf/Nip7/Rpusd2/Rpf1/Dkc1/Gtpbp4/Heatr3/Emg1/Cul4b/Rbm34/Mrto4/Ddx56/Urb1/Krr1/Nol9/Rexo4/Rpp40/Bud23/Exosc7/Npm3/Dhx37/Fcf1/Nsun5/Rrs1/Utp14a/Wdr36/Ngdn/Rps13/Rps6/Cinp/Rpp30/Eif2a/Mterf3/Utp18/Rrp15/Rps5/Exosc5/Tma16/Naf1/Rpl14/Rpl35/Eif4a3/Pno1/Exosc10/Rps19bp1/Rrp9/Ddx52/Tbl3/Mak16/Airim/Mterf4/Ddx47/Noc2l/Surf6/Mrm3/Pop7/Riok2/Eri3/Rps21/Rps8/Rpl7l1/Tsr2/Rpl7/Drosha/Rpsa/Srfbp1/Rrp8/Ltv1/Rps17/Exosc8/Rpusd4/Utp3/Rps16/Utp11/Znhit3/Riox2/Fastkd2/Rcc1l/Nup88/Rps15a/Rpp38/Nudt16/Ddx54/Ddx51/Dhx29/Isg20l2/Cdkn2a/Utp25/Rps27rt/Dimt1/Lsg1/Nvl/Pes1/Imp4/Pelp1/Tsc1/Mrm2/Eri2/Heatr1/Nol6/Rps9/Mphosph6/Slx9/Noa1/Bysl/Rps27a/Nol7/Zfp622/Sart1/Ercc2/Mrpl10/Cul4a/Rpl35a/Rpl27/Afg2b/Nom1/Nsa2/Rps28/Mettl17/Nob1/Wdr18/Rps23/Ddx31/Nol11/Znhit6/Las1l/Tfb2m/Rpf2/Pih1d1/Abt1/Rps12/Utp6/Pin4/Isg20/Utp14b/Mettl18/Usp16/Mpv17l2/Zfp593/Rbm10/Rsl24d1/Rpl7a/Rpl11/Mdn1/Rbfa/Snu13/Rps25/Rps24/Efl1/Usp36/Sde2/Ddx49/Fbll1/Utp23/Prkdc/Pop5/Rpl38/Zcchc4/Rpl5/Ipo9/Rpl24/Rps19/Eif1a/Wbp11/Mrm1/Kri1/Pwp1/Rpl26/Tfb1m/Mrps7/Rps27l/Ythdf2/Frg1/Mettl15/Tent4b/Pten/Rps15/Rps27/Rps11/Eif5b/Nop9/LTO1/Fau/Ddx28/Eral1/Lsm6/Rrp1/Pop4/Rrn3/Nsun3/Mettl5/Pih1d2/Afg2a/Rpusd1/Tsr3/Riok3/Rexo5/Rrp7a/Rps14/Xrcc5/Rexo1/Trmt2b/Mcat
GO:0006397 Srpk1/Srsf6/C1qbp/Ncl/Ddx20/Npm1/Hnrnpa2b1/Cstf1/Cpsf3/Wdr77/Mtrex/Dhx15/Sltm/Aqr/Prpf3/Nsun2/Sf3a3/Trmt6/Trmt2a/Rbm28/Ctnnbl1/Mettl16/Ddx23/Prpf4b/Pnpt1/Cdc5l/Gemin6/Rprd1b/Sf3b2/Ddx41/Tbrg4/Rbm39/Hspa8/Ppih/Magoh/Rrp1b/Ptcd2/Upf1/Srrt/Ddx39a/Luc7l2/Tra2a/Snrpb/Rnmt/Thoc3/Prpf38b/Lsm2/Srsf2/Ttf2/Prmt7/Pus1/Prpf19/U2af2/Gemin5/Rpusd2/Bud31/Rbm17/Syncrip/Hnrnpm/Sf3a2/Scaf11/Hnrnpk/Dus3l/Snrpd3/Dhx16/Prpf31/Pabpc1/Slbp/Papola/Cirbp/Bcas2/Ppp1r8/Smn1/Pus7/Hnrnpl/Cpsf1/Ppil3/Khsrp/Sf3b3/Ccnb1/Ptbp1/Rbm25/Cpsf6/Cpsf4/Hnrnpu/Rnf113a2/Sf3b4/Prpf40a/Zfp64/Dhx9/Srsf1/Aar2/Ncbp2/Cdc73/Snrnp200/Arl6ip4/Sf3b1/Sf3a1/Isy1/Ppil1/Eif4a3/Tcerg1/Tardbp/Cdk12/Exosc10/Ncbp1/Fastkd5/Ybx1/Trub2/Thrap3/Snrpb2/Mettl3/Prpf4/Snrnp70/Rbm38/Lsm3/Sf1/Upf3b/Mbnl2/Trmt10c/Ddx5/Raly/Ddx1/Lsm11/Snrpe/Dbr1/Tut1/Fra10ac1/Acin1/Ddx47/U2af1/Cstf2t/Srrm1/Pnn/Prmt5/Sart3/Iws1/Trmt61a/Alyref/Gemin2/Snrnp48/Thoc1/Bud13/Usp39/Lsm8/Larp7/Pcf11/Rpusd4/Arglu1/Ythdc1/Sf3b5/Snrpa1/Zfp473/Dhx8/Rbpms/Smndc1/Ppwd1/Srsf10/Ppp4r2/Ivns1abp/Apobec1/Nol3/Prpf6/Safb/Arvcf/Mtpap/Clp1/Jmjd6/Rbm8a/Srsf7/Phf5a/Tsen54/Cmtr2/Snrpn/Smu1/Paxbp1/Grsf1/Rbm15/Lsm1/Pus7l/Safb2/Hnrnpll/Zpr1/Ubl5/Luc7l3/Cwc22/Cactin/Lsm4/Setx/Syf2/Ssu72/Snrpd1/Prkaca/Hsf1/Rbmxl1/Tsen15/Khdc4/Rbpms2/Snrpf/Obi1/Gemin4/Cpsf2/Snrnp40/Csdc2/Sfpq/Usp4/Mfap1b/Rprd2/Sart1/Trub1/Xab2/Khdrbs3/Celf2/Pabpn1/Pcbp1/Prpf39/Scaf4/Celf5/Snrpg/Tsen2/Sugp1/Hnrnph1/Ecd/Srek1/Slu7/Nono/Tent2/Snrpd2/Rsrp1/Prpf8/Armc7/Adar/Thoc2/Xrn2/Hnrnpf/Cenatac/Tra2b/Hnrnpa3/Rbfox3/Hnrnpa1/Prmt9/Qki/Rbmx2/Plcb1/Snrpc/Clns1a/Sf3b6/Cwf19l1/Nova2/Kin/Rbm14/Txnl4a/Lsm5/Hmx2/Srrm2/Khdrbs2/Thoc5/Srrm4/Aff2/Rbm7/Rbm10/Snrpa/Cnot6l/U2af1l4/Zranb2/Rbm3/Yju2/Zcrb1/Mbnl1/Srsf3/Gpatch1/Celf3/Snu13/Rnpc3/Adarb2/Ahcyl1/Zcchc8/Zfp36l1/Clasrp/Rbm42/Rbm20/Nrde2/Sympk/Habp4/Ppie/Cstf3/Prkrip1/Zbtb7a/Thoc6/Ppil2/Tssc4/Pcbp4/Txnl4a/Sde2/Mblac1/Dhx38/Cpsf7/Nudt21/Snrnp25/Gcfc2/Cdc40/Mbnl3/Adarb1/Cstf2/Rbm11/Hdac7/Zrsr2/Srsf9/Prpf18/Strap/Lsm7/Phrf1/Zmat5/Scaf1/Rbm24/Cwc25/Luc7l/Crnkl1/Rnf113a1/Alkbh5/Eif1/Sfswap/Celf1/Celf6/Ern1/Nova1/Wbp11/Rbm6/Srpk3/Ncbp3/Cdk9/Wdr33/Tsen34/Rnps1/Wdr83/Lsm10/Ddx39b/Papolg/Scaf8/Dyrk1a/Ddx46/Zfp830/Hnrnpc/Mfap1a/Rbm22/Brdt/Prpf38a/Rbm48/Cript/Akap17b/Frg1/Akap8l/Tent4b/Fam50a/Sap18b/Ramac/Snip1/Arb2a/Aicda/Ilf3/Zc3h13/Dazap1/Lsm6/Pde12/Sap18/Rest/Pan3/Hnrnpul1/Alyref2/Rbfox1/Nup98/Dcps/Adam3/Leo1/Coil/Tent4a/Htatsf1/Nsrp1/Plrg1/Supt6/Zfp326/Malat1/Rbm5/Gemin7/Zmat2/Pan2/Cd2bp2/Cdk13/Slirp/Ptbp3/Rbm47/Rsrc1/Fmr1/Mettl14/Srsf12/Puf60/Lgals3
GO:0006364                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                        Nhp2/Ddx27/Ncl/Wdr3/Npm1/Nop2/Ddx21/Utp4/Pdcd11/Ddx18/Bms1/Mtrex/Lyar/Wdr46/Wdr12/Nol8/Ftsj3/Nop14/Utp15/Rps7/Pa2g4/Riok1/Nop10/Esf1/Wdr55/Ppan/Imp3/Exosc3/Eif6/Tsr1/Mettl16/Bop1/Gar1/Ebna1bp2/Nol10/Taf9/Dcaf13/Mphosph10/Suv39h1/Pwp2/Brix1/Wdr43/Exosc2/Rrp1b/Ddx10/Rcl1/Exosc1/Nat10/Fbl/Utp20/Wdr75/Wdr74/Pak1ip1/Eri1/Dis3/Rpusd2/Rpf1/Dkc1/Gtpbp4/Emg1/Rbm34/Mrto4/Ddx56/Urb1/Krr1/Nol9/Rexo4/Rpp40/Bud23/Exosc7/Npm3/Dhx37/Fcf1/Nsun5/Rrs1/Utp14a/Wdr36/Ngdn/Rps6/Rpp30/Utp18/Rrp15/Exosc5/Naf1/Rpl14/Rpl35/Eif4a3/Exosc10/Rrp9/Ddx52/Tbl3/Mak16/Mterf4/Ddx47/Mrm3/Pop7/Riok2/Eri3/Rps21/Rps8/Rpl7l1/Tsr2/Rpl7/Drosha/Srfbp1/Rrp8/Rps17/Exosc8/Rpusd4/Utp3/Rps16/Utp11/Znhit3/Rpp38/Nudt16/Ddx54/Ddx51/Cdkn2a/Utp25/Rps27rt/Dimt1/Nvl/Pes1/Imp4/Pelp1/Mrm2/Eri2/Heatr1/Nol6/Mphosph6/Slx9/Bysl/Nol7/Sart1/Ercc2/Rpl35a/Rpl27/Nsa2/Rps28/Nob1/Wdr18/Nol11/Znhit6/Las1l/Tfb2m/Rpf2/Pih1d1/Abt1/Utp6/Pin4/Isg20/Utp14b/Mettl18/Rpl7a/Rpl11/Rbfa/Snu13/Rps25/Rps24/Usp36/Sde2/Ddx49/Fbll1/Utp23/Prkdc/Pop5/Zcchc4/Rpl5/Rps19/Wbp11/Mrm1/Kri1/Pwp1/Rpl26/Tfb1m/Ythdf2/Frg1/Mettl15/Tent4b/Rps15/Rps27/Nop9/Lsm6/Rrp1/Pop4/Nsun3/Mettl5/Pih1d2/Rpusd1/Tsr3/Riok3/Rexo5/Rrp7a/Rps14/Rexo1/Trmt2b
GO:0022618                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                         Srpk1/Srsf6/Ddx20/Nop2/Wdr77/Eif5/Eif3j2/Eif3g/Prpf3/Sf3a3/Ruvbl1/Denr/Eif6/Bop1/Prpf4b/Taf9/Gemin6/Pwp2/Brix1/Eif3b/Ago1/Luc7l2/Snrpb/Eif2s2/Lsm2/Prmt7/Prpf19/U2af2/Gemin5/Tarbp2/Dkc1/Sf3a2/Scaf11/Snrpd3/Eif3a/Mrto4/Prpf31/Smn1/Atr/Eif3l/Rrs1/Eif3e/Rps6/Dhx9/Srsf1/Aar2/Rps5/Snrnp200/Naf1/Sf3b1/Sf3a1/Ago4/Isy1/Ncbp1/Mcts2/Hsp90ab1/Hsp90aa1/Shq1/Snrpe/Eif3j1/Prmt5/Sart3/Gemin2/Nufip1/Rpsa/Usp39/Ythdc1/Znhit3/Eif2s3y/Fastkd2/Srsf10/Nol3/Prpf6/Clp1/Eif3i/Ruvbl2/Dhx29/Nopchap1/Nvl/Luc7l3/Mrm2/Eif3c/Lsm4/Eif3m/Setx/Snrpd1/Khdc4/Snrpf/Gemin4/Eif3d/Eif3h/Usp4/Xab2/Celf2/Prpf39/Celf5/Snrpg/Rps28/Mettl17/Znhit6/Eif2s3x/Atm/Slu7/Rpf2/Snrpd2/Rsrp1/Prpf8/Pih1d1/Abt1/Adar/Snrpc/Clns1a/Ckap5/Yju2/Rpl11/Mbnl1/Mdn1/Celf3/Snu13/Rps25/Prkra/Tssc4/Gcfc2/Prkdc/Zrsr2/Prpf18/Rpl38/Strap/Rpl5/Luc7l/Crnkl1/Rpl24/Ptges3/Rps19/Sfswap/Celf1/Celf6/Srpk3/Ddx39b/Ddx46/Mrps7/Rps27l/Rps15/Rps27/Fau/Ddx28/Eral1/Eif2d/Ago3/Pih1d2/Coil/Ago2/Rpl13a/Htatsf1/Rrp7a/Rps14/Rbm5/Xrcc5/Gemin7/Eif3k/Mcat/Eif3f/Srsf12/Puf60
GO:0016072                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                Nhp2/Ddx27/Ncl/Polr1b/Wdr3/Npm1/Nop2/Ddx21/Utp4/Pdcd11/Ddx18/Bms1/Mtrex/Lyar/Wdr46/Wdr12/Nol8/Ftsj3/Nop14/Utp15/Rps7/Pa2g4/Riok1/Nop10/Esf1/Wdr55/Ppan/Imp3/Exosc3/Eif6/Tsr1/Mettl16/Bop1/Gar1/Ebna1bp2/Nol10/Taf9/Dcaf13/Mphosph10/Suv39h1/Pwp2/Polr1d/Brix1/Wdr43/Exosc2/Rrp1b/Ddx10/Tcof1/Rcl1/Exosc1/Nat10/Fbl/Utp20/Polr1g/Wdr75/Slfn8/Wdr74/Pak1ip1/Eri1/Dis3/Rpusd2/Rpf1/Dkc1/Gtpbp4/Emg1/Ddx11/Pelo/Rbm34/Mrto4/Cavin1/Ddx56/Urb1/Krr1/Nol9/Rexo4/Rpp40/Bud23/Exosc7/Npm3/Dhx37/Fcf1/Nsun5/Rrs1/Utp14a/Wdr36/Macroh2a1/Ngdn/Rps6/Rpp30/Utp18/Rrp15/Exosc5/Naf1/Rpl14/Rpl35/Eif4a3/Exosc10/Rrp9/Ddx52/Polr1a/Tbl3/Mak16/Mterf4/Ddx47/Mrm3/Pop7/Riok2/Eri3/Rps21/Rps8/Rpl7l1/Tsr2/Rpl7/Myc/Drosha/Srfbp1/Rrp8/Rps17/Exosc8/Rpusd4/Utp3/Ippk/Rps16/Utp11/Znhit3/Ubtf/Polr1e/Rpp38/Nudt16/Ddx54/Ddx51/Cdkn2a/Utp25/Top1/Rps27rt/Dimt1/Nvl/Pes1/Imp4/Pelp1/Mrm2/Eri2/Heatr1/Nol6/Mphosph6/Polr2e/Slx9/Bysl/Mars1/Ang/Polr1f/Nol7/Sart1/Macroh2a2/Ercc2/Rpl35a/Rpl27/Nsa2/Rps28/Nob1/Wdr18/Nol11/Znhit6/Las1l/Tfb2m/Rpf2/Pih1d1/Abt1/Utp6/Pin4/Isg20/Utp14b/Trir/Taf1b/Mettl18/Slfn9/Rpl7a/Rpl11/Rbfa/Snu13/Rps25/Rps24/Usp36/Sde2/Ddx49/Fbll1/Utp23/Prkdc/Pop5/Zcchc4/Rpl5/Rps19/Xrn1/Wbp11/Mrm1/Kri1/Pwp1/Rpl26/Tfb1m/Ythdf2/Frg1/Mettl15/Tent4b/Trp53/Rps15/Rps27/Nop9/Lsm6/Rrp1/Pop4/Ercc3/Nsun3/Mettl5/Pih1d2/Rpusd1/Tsr3/Riok3/Rexo5/Rrp7a/Rps14/Rexo1/Trmt2b/Spin1
           Count
GO:0022613   401
GO:0042254   283
GO:0006397   397
GO:0006364   196
GO:0022618   164
GO:0016072   223</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb97"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb97-1"><a href="#cb97-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Dotplot </span></span>
<span id="cb97-2"><a href="#cb97-2" aria-hidden="true" tabindex="-1"></a><span class="fu">dotplot</span>(ego, <span class="at">showCategory=</span><span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="step6_func_analysis_files/figure-html/unnamed-chunk-18-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb98"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb98-1"><a href="#cb98-1" aria-hidden="true" tabindex="-1"></a>p_dot_ego <span class="ot">&lt;-</span> <span class="fu">dotplot</span>(ego, <span class="at">showCategory=</span><span class="dv">15</span>)</span>
<span id="cb98-2"><a href="#cb98-2" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"plots/plot15_GO_dotplot.png"</span>, <span class="at">plot =</span> p_dot_ego, <span class="at">width =</span> <span class="dv">7</span>, <span class="at">height =</span> <span class="dv">7</span>)</span>
<span id="cb98-3"><a href="#cb98-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"plots/plot15_GO_dotplot.svg"</span>, <span class="at">plot =</span> p_dot_ego, <span class="at">width =</span> <span class="dv">7</span>, <span class="at">height =</span> <span class="dv">7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb99"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb99-1"><a href="#cb99-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Add similarity matrix to the termsim slot of enrichment result</span></span>
<span id="cb99-2"><a href="#cb99-2" aria-hidden="true" tabindex="-1"></a>ego <span class="ot">&lt;-</span> enrichplot<span class="sc">::</span><span class="fu">pairwise_termsim</span>(ego)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb100"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb100-1"><a href="#cb100-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Enrichmap clusters the most significant (by padj) GO terms to visualize relationships between terms</span></span>
<span id="cb100-2"><a href="#cb100-2" aria-hidden="true" tabindex="-1"></a><span class="fu">emapplot</span>(ego, <span class="at">showCategory =</span> <span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: Using `size` aesthetic for lines was deprecated in ggplot2 3.4.0.
ℹ Please use `linewidth` instead.
ℹ The deprecated feature was likely used in the ggtangle package.
  Please report the issue to the authors.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="step6_func_analysis_files/figure-html/unnamed-chunk-21-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb102"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb102-1"><a href="#cb102-1" aria-hidden="true" tabindex="-1"></a><span class="do">## To color genes by log2 fold changes, we need to extract the log2 fold changes from our results table creating a named vector</span></span>
<span id="cb102-2"><a href="#cb102-2" aria-hidden="true" tabindex="-1"></a>OE_foldchanges <span class="ot">&lt;-</span> sigOE<span class="sc">$</span>log2FoldChange</span>
<span id="cb102-3"><a href="#cb102-3" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(OE_foldchanges) <span class="ot">&lt;-</span> sigOE<span class="sc">$</span>gene</span>
<span id="cb102-4"><a href="#cb102-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(OE_foldchanges)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ENSMUSG00000001056 ENSMUSG00000028312 ENSMUSG00000003779 ENSMUSG00000004865 
          2.489647           3.195255           4.661063           2.927990 
ENSMUSG00000015176 ENSMUSG00000016921 
          4.208603           2.033734 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb104"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb104-1"><a href="#cb104-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Cnetplot details the genes associated with one or more terms - by default gives the top 5 significant terms (by padj)</span></span>
<span id="cb104-2"><a href="#cb104-2" aria-hidden="true" tabindex="-1"></a><span class="fu">cnetplot</span>(ego, </span>
<span id="cb104-3"><a href="#cb104-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">showCategory =</span> <span class="dv">1</span>, </span>
<span id="cb104-4"><a href="#cb104-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">color.params=</span><span class="fu">list</span>(<span class="at">foldChange=</span>OE_foldchanges),</span>
<span id="cb104-5"><a href="#cb104-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">vertex.label.font=</span><span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: ggrepel: 207 unlabeled data points (too many overlaps). Consider
increasing max.overlaps</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="step6_func_analysis_files/figure-html/unnamed-chunk-23-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb106"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb106-1"><a href="#cb106-1" aria-hidden="true" tabindex="-1"></a><span class="do">## If some of the high fold changes are getting drowned out due to a large range, you could set a maximum fold change value</span></span>
<span id="cb106-2"><a href="#cb106-2" aria-hidden="true" tabindex="-1"></a>OE_foldchanges <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(OE_foldchanges <span class="sc">&gt;</span> <span class="dv">2</span>, <span class="dv">2</span>, OE_foldchanges)</span>
<span id="cb106-3"><a href="#cb106-3" aria-hidden="true" tabindex="-1"></a>OE_foldchanges <span class="ot">&lt;-</span> <span class="fu">ifelse</span>(OE_foldchanges <span class="sc">&lt;</span> <span class="sc">-</span><span class="dv">2</span>, <span class="sc">-</span><span class="dv">2</span>, OE_foldchanges)</span>
<span id="cb106-4"><a href="#cb106-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb106-5"><a href="#cb106-5" aria-hidden="true" tabindex="-1"></a><span class="fu">cnetplot</span>(ego, </span>
<span id="cb106-6"><a href="#cb106-6" aria-hidden="true" tabindex="-1"></a>         <span class="at">categorySize=</span><span class="st">"pvalue"</span>, </span>
<span id="cb106-7"><a href="#cb106-7" aria-hidden="true" tabindex="-1"></a>         <span class="at">foldChange=</span>OE_foldchanges, </span>
<span id="cb106-8"><a href="#cb106-8" aria-hidden="true" tabindex="-1"></a>         <span class="at">showCategory =</span> <span class="dv">1</span>, </span>
<span id="cb106-9"><a href="#cb106-9" aria-hidden="true" tabindex="-1"></a>         <span class="at">vertex.label.font=</span><span class="dv">6</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: ggrepel: 223 unlabeled data points (too many overlaps). Consider
increasing max.overlaps</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="step6_func_analysis_files/figure-html/unnamed-chunk-24-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb108"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb108-1"><a href="#cb108-1" aria-hidden="true" tabindex="-1"></a>p_cnet_ego <span class="ot">&lt;-</span> <span class="fu">cnetplot</span>(ego, </span>
<span id="cb108-2"><a href="#cb108-2" aria-hidden="true" tabindex="-1"></a>         <span class="at">categorySize=</span><span class="st">"pvalue"</span>, </span>
<span id="cb108-3"><a href="#cb108-3" aria-hidden="true" tabindex="-1"></a>         <span class="at">foldChange=</span>OE_foldchanges, </span>
<span id="cb108-4"><a href="#cb108-4" aria-hidden="true" tabindex="-1"></a>         <span class="at">showCategory =</span> <span class="dv">1</span>, </span>
<span id="cb108-5"><a href="#cb108-5" aria-hidden="true" tabindex="-1"></a>         <span class="at">vertex.label.font=</span><span class="dv">6</span>)</span>
<span id="cb108-6"><a href="#cb108-6" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"plots/plot18_cnet_1cat_lfc.png"</span>, <span class="at">plot =</span> p_cnet_ego, <span class="at">width =</span> <span class="dv">7</span>, <span class="at">height =</span> <span class="dv">7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: ggrepel: 105 unlabeled data points (too many overlaps). Consider
increasing max.overlaps</code></pre>
</div>
<div class="sourceCode cell-code" id="cb110"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb110-1"><a href="#cb110-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"plots/plot18_cnet_1cat_lfc.svg"</span>, <span class="at">plot =</span> p_cnet_ego, <span class="at">width =</span> <span class="dv">7</span>, <span class="at">height =</span> <span class="dv">7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: ggrepel: 104 unlabeled data points (too many overlaps). Consider
increasing max.overlaps</code></pre>
</div>
</div>
</section>
<section id="httpshbctraining.github.iointro-to-dgelessons11_fa_functional_class_scoring.html" class="level2">
<h2 class="anchored" data-anchor-id="httpshbctraining.github.iointro-to-dgelessons11_fa_functional_class_scoring.html">https://hbctraining.github.io/Intro-to-DGE/lessons/11_FA_functional_class_scoring.html</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb112"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb112-1"><a href="#cb112-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Remove any NA values (reduces the data by quite a bit)</span></span>
<span id="cb112-2"><a href="#cb112-2" aria-hidden="true" tabindex="-1"></a>res_entrez <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(res_ids, entrezid <span class="sc">!=</span> <span class="st">"NA"</span>)</span>
<span id="cb112-3"><a href="#cb112-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_entrez)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 12
  gene   baseMean log2FoldChange  lfcSE pvalue  padj threshold symbol genelabels
  &lt;chr&gt;     &lt;dbl&gt;          &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;lgl&gt;     &lt;chr&gt;  &lt;chr&gt;     
1 ENSMU…    2198.           2.49 0.0625      0     0 TRUE      Nhp2   Nhp2      
2 ENSMU…    1186.           3.20 0.0829      0     0 TRUE      Smc2   Smc2      
3 ENSMU…    1362.           4.66 0.0951      0     0 TRUE      Kif20a Kif20a    
4 ENSMU…    1272.           2.93 0.0748      0     0 TRUE      Srpk1  Srpk1     
5 ENSMU…    4374.           4.21 0.0917      0     0 TRUE      Nolc1  Nolc1     
6 ENSMU…    3387.           2.03 0.0537      0     0 TRUE      Srsf6  Srsf6     
# ℹ 3 more variables: gene_name &lt;chr&gt;, entrezid &lt;int&gt;, gene_biotype &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb114"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb114-1"><a href="#cb114-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(res_entrez)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 15977</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb116"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb116-1"><a href="#cb116-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Remove any Entrez duplicates</span></span>
<span id="cb116-2"><a href="#cb116-2" aria-hidden="true" tabindex="-1"></a>res_entrez <span class="ot">&lt;-</span> res_entrez[<span class="fu">which</span>(<span class="fu">duplicated</span>(res_entrez<span class="sc">$</span>entrezid) <span class="sc">==</span> F), ]</span>
<span id="cb116-3"><a href="#cb116-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(res_entrez)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code># A tibble: 6 × 12
  gene   baseMean log2FoldChange  lfcSE pvalue  padj threshold symbol genelabels
  &lt;chr&gt;     &lt;dbl&gt;          &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;lgl&gt;     &lt;chr&gt;  &lt;chr&gt;     
1 ENSMU…    2198.           2.49 0.0625      0     0 TRUE      Nhp2   Nhp2      
2 ENSMU…    1186.           3.20 0.0829      0     0 TRUE      Smc2   Smc2      
3 ENSMU…    1362.           4.66 0.0951      0     0 TRUE      Kif20a Kif20a    
4 ENSMU…    1272.           2.93 0.0748      0     0 TRUE      Srpk1  Srpk1     
5 ENSMU…    4374.           4.21 0.0917      0     0 TRUE      Nolc1  Nolc1     
6 ENSMU…    3387.           2.03 0.0537      0     0 TRUE      Srsf6  Srsf6     
# ℹ 3 more variables: gene_name &lt;chr&gt;, entrezid &lt;int&gt;, gene_biotype &lt;chr&gt;</code></pre>
</div>
<div class="sourceCode cell-code" id="cb118"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb118-1"><a href="#cb118-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(res_entrez)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 15969</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb120"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb120-1"><a href="#cb120-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Extract the foldchanges</span></span>
<span id="cb120-2"><a href="#cb120-2" aria-hidden="true" tabindex="-1"></a>foldchanges <span class="ot">&lt;-</span> res_entrez<span class="sc">$</span>log2FoldChange</span>
<span id="cb120-3"><a href="#cb120-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-4"><a href="#cb120-4" aria-hidden="true" tabindex="-1"></a><span class="do">## Name each fold change with the corresponding Entrez ID</span></span>
<span id="cb120-5"><a href="#cb120-5" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(foldchanges) <span class="ot">&lt;-</span> res_entrez<span class="sc">$</span>entrezid</span>
<span id="cb120-6"><a href="#cb120-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-7"><a href="#cb120-7" aria-hidden="true" tabindex="-1"></a><span class="do">## Sort fold changes in decreasing order</span></span>
<span id="cb120-8"><a href="#cb120-8" aria-hidden="true" tabindex="-1"></a>foldchanges <span class="ot">&lt;-</span> <span class="fu">sort</span>(foldchanges, <span class="at">decreasing =</span> <span class="cn">TRUE</span>)</span>
<span id="cb120-9"><a href="#cb120-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb120-10"><a href="#cb120-10" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(foldchanges)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>   56746   380684   110542    14412    15404    13038 
16.20879 12.71670 12.27802 12.17475 11.77999 11.55211 </code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb122"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb122-1"><a href="#cb122-1" aria-hidden="true" tabindex="-1"></a><span class="fu">set.seed</span>(<span class="dv">123456</span>)</span>
<span id="cb122-2"><a href="#cb122-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb122-3"><a href="#cb122-3" aria-hidden="true" tabindex="-1"></a><span class="do">## GSEA using gene sets from KEGG pathways</span></span>
<span id="cb122-4"><a href="#cb122-4" aria-hidden="true" tabindex="-1"></a>gseaKEGG <span class="ot">&lt;-</span> <span class="fu">gseKEGG</span>(<span class="at">geneList =</span> foldchanges, <span class="co"># ordered named vector of fold changes (Entrez IDs are the associated names)</span></span>
<span id="cb122-5"><a href="#cb122-5" aria-hidden="true" tabindex="-1"></a>              <span class="at">organism =</span> <span class="st">"mmu"</span>, <span class="co"># supported organisms listed below</span></span>
<span id="cb122-6"><a href="#cb122-6" aria-hidden="true" tabindex="-1"></a>              <span class="at">minGSSize =</span> <span class="dv">20</span>, <span class="co"># minimum gene set size (# genes in set) - change to test more sets or recover sets with fewer # genes</span></span>
<span id="cb122-7"><a href="#cb122-7" aria-hidden="true" tabindex="-1"></a>              <span class="at">pvalueCutoff =</span> <span class="fl">0.05</span>, <span class="co"># padj cutoff value</span></span>
<span id="cb122-8"><a href="#cb122-8" aria-hidden="true" tabindex="-1"></a>              <span class="at">verbose =</span> <span class="cn">FALSE</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading KEGG annotation online: "https://rest.kegg.jp/link/mmu/pathway"...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Reading KEGG annotation online: "https://rest.kegg.jp/list/pathway/mmu"...</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in preparePathwaysAndStats(pathways, stats, minSize, maxSize, gseaParam, : There are ties in the preranked stats (0.54% of the list).
The order of those tied genes will be arbitrary, which may produce unexpected results.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in fgseaMultilevel(pathways = pathways, stats = stats, minSize =
minSize, : There were 12 pathways for which P-values were not calculated
properly due to unbalanced (positive and negative) gene-level statistic values.
For such pathways pval, padj, NES, log2err are set to NA. You can try to
increase the value of the argument nPermSimple (for example set it nPermSimple
= 10000)</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in fgseaMultilevel(pathways = pathways, stats = stats, minSize =
minSize, : For some of the pathways the P-values were likely overestimated. For
such pathways log2err is set to NA.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning in fgseaMultilevel(pathways = pathways, stats = stats, minSize =
minSize, : For some pathways, in reality P-values are less than 1e-10. You can
set the `eps` argument to zero for better estimation.</code></pre>
</div>
<div class="sourceCode cell-code" id="cb129"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb129-1"><a href="#cb129-1" aria-hidden="true" tabindex="-1"></a><span class="fu">class</span>(gseaKEGG)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "gseaResult"
attr(,"package")
[1] "DOSE"</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb131"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb131-1"><a href="#cb131-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Extract the GSEA results</span></span>
<span id="cb131-2"><a href="#cb131-2" aria-hidden="true" tabindex="-1"></a>gseaKEGG_results <span class="ot">&lt;-</span> gseaKEGG<span class="sc">@</span>result</span>
<span id="cb131-3"><a href="#cb131-3" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(gseaKEGG_results)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>               ID                             Description setSize
mmu03030 mmu03030                         DNA replication      35
mmu04360 mmu04360                           Axon guidance     174
mmu04080 mmu04080 Neuroactive ligand-receptor interaction     185
mmu04517 mmu04517                      IgSF CAM signaling     269
mmu03460 mmu03460                  Fanconi anemia pathway      51
mmu03440 mmu03440                Homologous recombination      40
         enrichmentScore       NES       pvalue     p.adjust       qvalue rank
mmu03030       0.8250980  2.891843 1.000000e-10 7.575000e-09 5.421053e-09 2411
mmu04360      -0.6850729 -1.567544 1.000000e-10 7.575000e-09 5.421053e-09 3340
mmu04080      -0.6828048 -1.565159 1.000000e-10 7.575000e-09 5.421053e-09 3059
mmu04517      -0.6689996 -1.541695 1.000000e-10 7.575000e-09 5.421053e-09 2252
mmu03460       0.7143593  2.758993 1.715340e-10 1.039496e-08 7.439159e-09 2671
mmu03440       0.7134202  2.604239 2.090033e-08 1.055466e-06 7.553451e-07 2614
                           leading_edge
mmu03030 tags=91%, list=15%, signal=78%
mmu04360 tags=49%, list=21%, signal=39%
mmu04080 tags=50%, list=19%, signal=41%
mmu04517 tags=32%, list=14%, signal=28%
mmu03460 tags=63%, list=17%, signal=52%
mmu03440 tags=72%, list=16%, signal=61%
                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     core_enrichment
mmu03030                                                                                                                                                                                                                                                                                                                                                                                          17218/18973/17220/19075/18968/17215/17216/106344/18974/18969/19891/17217/14156/18971/327762/16881/19076/18972/17219/68240/19819/69263/381760/72151/19718/59001/19687/67967/69724/68209/68275/67153
mmu04360                                        13641/12168/13639/20353/12934/259302/26417/13836/20562/13837/20130/171171/319757/69632/12767/22066/107449/13643/13640/319713/20347/170758/13637/18762/18208/226251/20357/107448/18209/12323/20350/26456/19059/20564/18007/93737/20360/14365/22253/57764/231148/13800/19876/13841/20349/192198/17295/65254/13642/11854/20356/93742/20423/214230/20361/22418/53972/117606/210801/241656/20315/22063/235611/243743/12167/13848/20563/20346/13844/13839/117600/108151/12162/98932/270190/20348/13176/13838/214968/241568/14677/244058/18481/268902/20358
mmu04080 12672/14293/21336/21833/12273/114229/229214/19204/18166/14419/18436/15465/228139/11444/233571/12671/12267/74191/68039/18440/14289/19214/19216/140795/11553/19228/94226/14394/20604/12310/18155/19217/15558/110637/30878/14811/18430/14395/18389/243764/19773/14812/14658/14803/13609/12424/14809/12062/381677/14600/60531/14810/110886/320343/23936/12921/14407/15559/11535/15551/14745/18619/14406/14401/242425/14805/11548/20606/12801/13617/11550/21834/242443/13610/53623/109648/14396/14799/14405/75552/54393/14806/14804/14802/14807/14062/11539/11438/108071/14402/11517/14800/13618
mmu04517                              18208/22142/243547/53857/50791/107448/268932/116837/66797/54631/17136/71740/83964/12927/17196/26904/11735/110862/22153/69524/68507/20564/16542/93737/67374/22253/58998/16416/245446/19876/319504/327814/12805/71960/76965/244654/93742/21872/67703/331461/16492/214230/13823/243499/16485/210801/241656/269116/16490/19268/192167/16536/243743/22152/63955/18751/20563/16456/19094/17475/117600/170643/386750/98932/17968/260299/239857/13176/19266/20742/245450/13052/94332/18213/245537/26414/241568/272381/18481/268902/12558/17967/16494/17756/19283/73710
mmu03460                                                                                                                                                                                                                                                                                                                                                                             19361/237211/223970/12189/237911/67196/12144/211651/268465/14087/69928/230484/19891/233826/100040608/101831/208836/68240/12190/80905/104806/14088/18861/67030/71718/56210/245000/17350/68275/52864/114714/27015
mmu03440                                                                                                                                                                                                                                                                                                                                                                                                      19361/12189/12021/237911/623474/12144/74335/268465/235559/19891/19363/233826/18971/19366/70681/18972/68240/12190/381760/11920/19360/17535/57434/225182/67967/210766/68275/27354/114714</code></pre>
</div>
<div class="sourceCode cell-code" id="cb133"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb133-1"><a href="#cb133-1" aria-hidden="true" tabindex="-1"></a><span class="fu">nrow</span>(gseaKEGG_results)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] 70</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb135"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb135-1"><a href="#cb135-1" aria-hidden="true" tabindex="-1"></a><span class="do">## Plot the GSEA plot for a single enriched pathway</span></span>
<span id="cb135-2"><a href="#cb135-2" aria-hidden="true" tabindex="-1"></a><span class="fu">gseaplot</span>(gseaKEGG, <span class="at">geneSetID =</span> <span class="st">'mmu03030'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: `aes_()` was deprecated in ggplot2 3.0.0.
ℹ Please use tidy evaluation idioms with `aes()`
ℹ The deprecated feature was likely used in the enrichplot package.
  Please report the issue at
  &lt;https://github.com/GuangchuangYu/enrichplot/issues&gt;.</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="step6_func_analysis_files/figure-html/unnamed-chunk-31-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="httpsyulab-smu.topbiomedical-knowledge-mining-bookenrichplot.html" class="level2">
<h2 class="anchored" data-anchor-id="httpsyulab-smu.topbiomedical-knowledge-mining-bookenrichplot.html">https://yulab-smu.top/biomedical-knowledge-mining-book/enrichplot.html</h2>
<div class="cell">
<div class="sourceCode cell-code" id="cb137"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb137-1"><a href="#cb137-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(enrichplot)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: package 'enrichplot' was built under R version 4.4.2</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>enrichplot v1.26.6 Learn more at https://yulab-smu.top/contribution-knowledge-mining/

Please cite:

Guangchuang Yu. Gene Ontology Semantic Similarity Analysis Using
GOSemSim. In: Kidder B. (eds) Stem Cell Transcriptional Networks.
Methods in Molecular Biology. 2020, 2117:207-215. Humana, New York, NY.</code></pre>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb140"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb140-1"><a href="#cb140-1" aria-hidden="true" tabindex="-1"></a><span class="co"># upsetplot is an alternative to cnetplot for visualizing the complex association between genes and gene sets. It emphasizes the gene overlapping among different gene sets.</span></span>
<span id="cb140-2"><a href="#cb140-2" aria-hidden="true" tabindex="-1"></a><span class="fu">upsetplot</span>(ego)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="step6_func_analysis_files/figure-html/unnamed-chunk-33-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb141"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb141-1"><a href="#cb141-1" aria-hidden="true" tabindex="-1"></a><span class="co"># For GSEA result, it will plot the fold change distributions of different categories (e.g. unique to pathway, overlaps among different pathways).</span></span>
<span id="cb141-2"><a href="#cb141-2" aria-hidden="true" tabindex="-1"></a><span class="fu">upsetplot</span>(gseaKEGG)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="step6_func_analysis_files/figure-html/unnamed-chunk-34-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb142"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb142-1"><a href="#cb142-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ridgeplot</span>(gseaKEGG, <span class="at">showCategory=</span><span class="dv">15</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Warning: `aes_string()` was deprecated in ggplot2 3.0.0.
ℹ Please use tidy evaluation idioms with `aes()`.
ℹ See also `vignette("ggplot2-in-packages")` for more information.
ℹ The deprecated feature was likely used in the enrichplot package.
  Please report the issue at
  &lt;https://github.com/GuangchuangYu/enrichplot/issues&gt;.</code></pre>
</div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 1</code></pre>
</div>
<div class="cell-output-display">
<div>
<figure class="figure">
<p><img src="step6_func_analysis_files/figure-html/unnamed-chunk-35-1.png" class="img-fluid figure-img" width="672"></p>
</figure>
</div>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb145"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb145-1"><a href="#cb145-1" aria-hidden="true" tabindex="-1"></a><span class="co"># ridgeplot will visualize expression distributions of core enriched genes for GSEA enriched categories. It helps users to interpret up/down-regulated pathways.</span></span>
<span id="cb145-2"><a href="#cb145-2" aria-hidden="true" tabindex="-1"></a>p_ridge_kegg <span class="ot">&lt;-</span> <span class="fu">ridgeplot</span>(gseaKEGG, <span class="at">showCategory=</span><span class="dv">15</span>)</span>
<span id="cb145-3"><a href="#cb145-3" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"plots/plot22_ridge_kegg.png"</span>, <span class="at">plot =</span> p_ridge_kegg, <span class="at">width =</span> <span class="dv">7</span>, <span class="at">height =</span> <span class="dv">7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 1</code></pre>
</div>
<div class="sourceCode cell-code" id="cb147"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb147-1"><a href="#cb147-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">"plots/plot22_ridge_kegg.svg"</span>, <span class="at">plot =</span> p_ridge_kegg, <span class="at">width =</span> <span class="dv">7</span>, <span class="at">height =</span> <span class="dv">7</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stderr">
<pre><code>Picking joint bandwidth of 1</code></pre>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
  window.document.addEventListener("DOMContentLoaded", function (event) {
    const icon = "";
    const anchorJS = new window.AnchorJS();
    anchorJS.options = {
      placement: 'right',
      icon: icon
    };
    anchorJS.add('.anchored');
    const isCodeAnnotation = (el) => {
      for (const clz of el.classList) {
        if (clz.startsWith('code-annotation-')) {                     
          return true;
        }
      }
      return false;
    }
    const onCopySuccess = function(e) {
      // button target
      const button = e.trigger;
      // don't keep focus
      button.blur();
      // flash "checked"
      button.classList.add('code-copy-button-checked');
      var currentTitle = button.getAttribute("title");
      button.setAttribute("title", "Copied!");
      let tooltip;
      if (window.bootstrap) {
        button.setAttribute("data-bs-toggle", "tooltip");
        button.setAttribute("data-bs-placement", "left");
        button.setAttribute("data-bs-title", "Copied!");
        tooltip = new bootstrap.Tooltip(button, 
          { trigger: "manual", 
            customClass: "code-copy-button-tooltip",
            offset: [0, -8]});
        tooltip.show();    
      }
      setTimeout(function() {
        if (tooltip) {
          tooltip.hide();
          button.removeAttribute("data-bs-title");
          button.removeAttribute("data-bs-toggle");
          button.removeAttribute("data-bs-placement");
        }
        button.setAttribute("title", currentTitle);
        button.classList.remove('code-copy-button-checked');
      }, 1000);
      // clear code selection
      e.clearSelection();
    }
    const getTextToCopy = function(trigger) {
        const codeEl = trigger.previousElementSibling.cloneNode(true);
        for (const childEl of codeEl.children) {
          if (isCodeAnnotation(childEl)) {
            childEl.remove();
          }
        }
        return codeEl.innerText;
    }
    const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
      text: getTextToCopy
    });
    clipboard.on('success', onCopySuccess);
    if (window.document.getElementById('quarto-embedded-source-code-modal')) {
      const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
        text: getTextToCopy,
        container: window.document.getElementById('quarto-embedded-source-code-modal')
      });
      clipboardModal.on('success', onCopySuccess);
    }
      var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
      var mailtoRegex = new RegExp(/^mailto:/);
        var filterRegex = new RegExp('/' + window.location.host + '/');
      var isInternal = (href) => {
          return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
      }
      // Inspect non-navigation links and adorn them if external
     var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
      for (var i=0; i<links.length; i++) {
        const link = links[i];
        if (!isInternal(link.href)) {
          // undo the damage that might have been done by quarto-nav.js in the case of
          // links that we want to consider external
          if (link.dataset.originalHref !== undefined) {
            link.href = link.dataset.originalHref;
          }
        }
      }
    function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
      const config = {
        allowHTML: true,
        maxWidth: 500,
        delay: 100,
        arrow: false,
        appendTo: function(el) {
            return el.parentElement;
        },
        interactive: true,
        interactiveBorder: 10,
        theme: 'quarto',
        placement: 'bottom-start',
      };
      if (contentFn) {
        config.content = contentFn;
      }
      if (onTriggerFn) {
        config.onTrigger = onTriggerFn;
      }
      if (onUntriggerFn) {
        config.onUntrigger = onUntriggerFn;
      }
      window.tippy(el, config); 
    }
    const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
    for (var i=0; i<noterefs.length; i++) {
      const ref = noterefs[i];
      tippyHover(ref, function() {
        // use id or data attribute instead here
        let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
        try { href = new URL(href).hash; } catch {}
        const id = href.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note) {
          return note.innerHTML;
        } else {
          return "";
        }
      });
    }
    const xrefs = window.document.querySelectorAll('a.quarto-xref');
    const processXRef = (id, note) => {
      // Strip column container classes
      const stripColumnClz = (el) => {
        el.classList.remove("page-full", "page-columns");
        if (el.children) {
          for (const child of el.children) {
            stripColumnClz(child);
          }
        }
      }
      stripColumnClz(note)
      if (id === null || id.startsWith('sec-')) {
        // Special case sections, only their first couple elements
        const container = document.createElement("div");
        if (note.children && note.children.length > 2) {
          container.appendChild(note.children[0].cloneNode(true));
          for (let i = 1; i < note.children.length; i++) {
            const child = note.children[i];
            if (child.tagName === "P" && child.innerText === "") {
              continue;
            } else {
              container.appendChild(child.cloneNode(true));
              break;
            }
          }
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(container);
          }
          return container.innerHTML
        } else {
          if (window.Quarto?.typesetMath) {
            window.Quarto.typesetMath(note);
          }
          return note.innerHTML;
        }
      } else {
        // Remove any anchor links if they are present
        const anchorLink = note.querySelector('a.anchorjs-link');
        if (anchorLink) {
          anchorLink.remove();
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        if (note.classList.contains("callout")) {
          return note.outerHTML;
        } else {
          return note.innerHTML;
        }
      }
    }
    for (var i=0; i<xrefs.length; i++) {
      const xref = xrefs[i];
      tippyHover(xref, undefined, function(instance) {
        instance.disable();
        let url = xref.getAttribute('href');
        let hash = undefined; 
        if (url.startsWith('#')) {
          hash = url;
        } else {
          try { hash = new URL(url).hash; } catch {}
        }
        if (hash) {
          const id = hash.replace(/^#\/?/, "");
          const note = window.document.getElementById(id);
          if (note !== null) {
            try {
              const html = processXRef(id, note.cloneNode(true));
              instance.setContent(html);
            } finally {
              instance.enable();
              instance.show();
            }
          } else {
            // See if we can fetch this
            fetch(url.split('#')[0])
            .then(res => res.text())
            .then(html => {
              const parser = new DOMParser();
              const htmlDoc = parser.parseFromString(html, "text/html");
              const note = htmlDoc.getElementById(id);
              if (note !== null) {
                const html = processXRef(id, note);
                instance.setContent(html);
              } 
            }).finally(() => {
              instance.enable();
              instance.show();
            });
          }
        } else {
          // See if we can fetch a full url (with no hash to target)
          // This is a special case and we should probably do some content thinning / targeting
          fetch(url)
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.querySelector('main.content');
            if (note !== null) {
              // This should only happen for chapter cross references
              // (since there is no id in the URL)
              // remove the first header
              if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
                note.children[0].remove();
              }
              const html = processXRef(null, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      }, function(instance) {
      });
    }
        let selectedAnnoteEl;
        const selectorForAnnotation = ( cell, annotation) => {
          let cellAttr = 'data-code-cell="' + cell + '"';
          let lineAttr = 'data-code-annotation="' +  annotation + '"';
          const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
          return selector;
        }
        const selectCodeLines = (annoteEl) => {
          const doc = window.document;
          const targetCell = annoteEl.getAttribute("data-target-cell");
          const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
          const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
          const lines = annoteSpan.getAttribute("data-code-lines").split(",");
          const lineIds = lines.map((line) => {
            return targetCell + "-" + line;
          })
          let top = null;
          let height = null;
          let parent = null;
          if (lineIds.length > 0) {
              //compute the position of the single el (top and bottom and make a div)
              const el = window.document.getElementById(lineIds[0]);
              top = el.offsetTop;
              height = el.offsetHeight;
              parent = el.parentElement.parentElement;
            if (lineIds.length > 1) {
              const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
              const bottom = lastEl.offsetTop + lastEl.offsetHeight;
              height = bottom - top;
            }
            if (top !== null && height !== null && parent !== null) {
              // cook up a div (if necessary) and position it 
              let div = window.document.getElementById("code-annotation-line-highlight");
              if (div === null) {
                div = window.document.createElement("div");
                div.setAttribute("id", "code-annotation-line-highlight");
                div.style.position = 'absolute';
                parent.appendChild(div);
              }
              div.style.top = top - 2 + "px";
              div.style.height = height + 4 + "px";
              div.style.left = 0;
              let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
              if (gutterDiv === null) {
                gutterDiv = window.document.createElement("div");
                gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
                gutterDiv.style.position = 'absolute';
                const codeCell = window.document.getElementById(targetCell);
                const gutter = codeCell.querySelector('.code-annotation-gutter');
                gutter.appendChild(gutterDiv);
              }
              gutterDiv.style.top = top - 2 + "px";
              gutterDiv.style.height = height + 4 + "px";
            }
            selectedAnnoteEl = annoteEl;
          }
        };
        const unselectCodeLines = () => {
          const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
          elementsIds.forEach((elId) => {
            const div = window.document.getElementById(elId);
            if (div) {
              div.remove();
            }
          });
          selectedAnnoteEl = undefined;
        };
          // Handle positioning of the toggle
      window.addEventListener(
        "resize",
        throttle(() => {
          elRect = undefined;
          if (selectedAnnoteEl) {
            selectCodeLines(selectedAnnoteEl);
          }
        }, 10)
      );
      function throttle(fn, ms) {
      let throttle = false;
      let timer;
        return (...args) => {
          if(!throttle) { // first call gets through
              fn.apply(this, args);
              throttle = true;
          } else { // all the others get throttled
              if(timer) clearTimeout(timer); // cancel #2
              timer = setTimeout(() => {
                fn.apply(this, args);
                timer = throttle = false;
              }, ms);
          }
        };
      }
        // Attach click handler to the DT
        const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
        for (const annoteDlNode of annoteDls) {
          annoteDlNode.addEventListener('click', (event) => {
            const clickedEl = event.target;
            if (clickedEl !== selectedAnnoteEl) {
              unselectCodeLines();
              const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
              if (activeEl) {
                activeEl.classList.remove('code-annotation-active');
              }
              selectCodeLines(clickedEl);
              clickedEl.classList.add('code-annotation-active');
            } else {
              // Unselect the line
              unselectCodeLines();
              clickedEl.classList.remove('code-annotation-active');
            }
          });
        }
    const findCites = (el) => {
      const parentEl = el.parentElement;
      if (parentEl) {
        const cites = parentEl.dataset.cites;
        if (cites) {
          return {
            el,
            cites: cites.split(' ')
          };
        } else {
          return findCites(el.parentElement)
        }
      } else {
        return undefined;
      }
    };
    var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
    for (var i=0; i<bibliorefs.length; i++) {
      const ref = bibliorefs[i];
      const citeInfo = findCites(ref);
      if (citeInfo) {
        tippyHover(citeInfo.el, function() {
          var popup = window.document.createElement('div');
          citeInfo.cites.forEach(function(cite) {
            var citeDiv = window.document.createElement('div');
            citeDiv.classList.add('hanging-indent');
            citeDiv.classList.add('csl-entry');
            var biblioDiv = window.document.getElementById('ref-' + cite);
            if (biblioDiv) {
              citeDiv.innerHTML = biblioDiv.innerHTML;
            }
            popup.appendChild(citeDiv);
          });
          return popup.innerHTML;
        });
      }
    }
  });
  </script>
</div> <!-- /content -->




</body></html>